Implement a repeated scheduling mechanism

FranzBusch opened the pull request:
Create a way to schedule repeated tasks

Motivation:
I think it would be nice to have a method to repeatedly schedule a task, this implementation just provides the basics to do so, but is definitely not the final solution. In my opinion, we should provide a method to schedule a task with at least these parameters:

- delay
- interval
- numberOfTimes

The other part I was thinking about is that with my current implementation the scheduling is done by an `EventLoop` but you could as well provide a method on `RepeatedTask` like `func begin(on: EventLoop)` which would allow it to be scheduled on multiple `EventLoop`s at a time, which I don't know if it's thread safe?

I just wanted to get early input if my approach is going in the right direction and if you have any additional requirements for such an implementation.

FranzBusch commented on Sources/NIO/EventLoop.swift:
```diff
@@ -50,6 +50,43 @@ public struct Scheduled<T> {
+            reschedule()
+        }
+    }
+    private let task: () throws -> Void
```
I was thinking about this as well and I think there should be 2 different options one where a task is scheduled in an interval and one where there is a fixed delay after each task. I think both cases could make sense depending on what the user actually wants to achieve

weissi commented:
@Lukasa yes, they do continue accepting work (that they'll never run). The issue is that outside of the event loop thread you can't query the lifecycle state. We could and should probably fix this though. We could have another variable (like `scheduledTasks`) that mirrors (parts of) the `lifecycleState` but is readable from any thread when holding `tasksLock`.

Lukasa commented:
So the main problem with them accepting work is that it means that they'll get stuck in a reference cycle and never be freed. Any scheduled task holds a future that fires on that event loop when the task is completed. This future keeps the loop alive. This is already pretty bad, but a repeated task essentially guarantees an event loop leak because it repeatedly reschedules work, ensuring there is always an event loop reference alive.

FranzBusch commented:
So this is blocked until we adopt the EventLoop to not be retained by scheduledTasks right?

helje5 commented:
Maybe I don't get the issue, but if I schedule a repeating timer, I most definitely would want the event loop to stay alive and not go away (and presumably my timer with it). No?

FranzBusch commented:
I think it is quite reasonable to only have the repeating task scheduled as long as the event loop is actually running since the task should not be responsible for the lifetime of the eventloop right?

Lukasa commented:
> Maybe I don't get the issue, but if I schedule a repeating timer, I most definitely would want the event loop to stay alive and not go away (and presumably my timer with it). No?

The problem is not that the event loop will go away, it's that the event loop won't tell the repeating task that it will never execute again. That is, if you do this:

```swift
loop.scheduleRepeatedTask(interval: .seconds(10)) { print("Still alive") }
sleep(1000)
loop.syncShutdownGracefully()
```

Then after 1000 seconds the repeating task will stop executing (as the loop is closed), but the repeating task will never be told. Worse, the loop will hold a reference to the scheduled task item, which holds a reference to the loop, causing the loop to be leaked.

This is not the end of the world in production applications because generally speaking they do not stop and start loops regularly, but it's very annoying in tests, as any use of repeating tasks will cause each test to leak the loops unless the user very aggressively cleans up their repeating tasks.

Put another way, in an ideal world the task would be told when the event loop has gone away, and any attempt to schedule new work would fail. Neither of those things happens in NIO today: when we shut the loop down we leave any pending tasks pending, and those pending tasks almost always keep the event loop alive.

I don't know that this blocks the proposal, as the problem already exists today with scheduled tasks, but repeating tasks take this possibility and make it an inevitability, which is not necessarily an improvement.

FranzBusch commented:
So would a fix be as simple as removing all scheduled task while shutting down?

Lukasa commented:
@FranzBusch That's not quite sufficient, because those shutdown tasks are likely to be necessary for some parts of cleanup. We probably need to design our way out of this, likely by using more than one task queue.

FranzBusch commented:
I just updated the PR it now includes methods to schedule a task with a fixedRate or fixedDelay. Furthermore, I removed the retain cycle created by the repeated task. Documentation, tests, and two methods which take a closure with a `Void` return parameter as still left to do.
@weissi  and @Lukasa can you have a quick look if the implementation seems good so far?

weissi commented:
@lukasa & @FranzBusch do we actually need `fixedRate`? If we have an overloaded event loop and this keeps firing it will lead to the EL becomig more overloaded.
Also I’m sure sure I like the names (`fixedRate` vs `fixedDelay`) too much. I think we could use long and descriptive names. But ideally we would only have what we call fixedDelay today.

FranzBusch commented:
@weissi For my personal project, I only used fixedRate as it was used for a polling mechanism and should fire every second even though the last invocation may still be processing. I get your point that we may overload the EL if the triggered task is being queued again and again, but I see use cases where fixed rate makes sense. 
Regarding the naming, I had a look at the netty implementation and went according to that, though I am open for better suggestions :)

weissi commented:
@FranzBusch makes sense. If you really want that, why don't you just use `eventloop.execute { myActualWork }; return eventloop.newSucceededFuture(value: ())` or similar to perform the other stuff? Then it will still fire every second even those `myActualWork` might still be going on?

FranzBusch commented:
@weissi Maybe, I don't get your example correctly but how would a call to `execute` repeatedly schedule a task and how would I cancel such a task sometimes in the future?

weissi commented:
@FranzBusch Sorry, I wasn't quite so clear: My assumption is that `.fixedDelay` and `.fixedRate` are the same if the work item that is scheduled doesn't take any time to execute (ie. succeeds the promise immediately). If that assumption is correct, you can just use `.fixedDelay` and _make_ the work item take no time but enqueueing the real work item using `el.execute`. Does that make sense?

weissi commented on Sources/NIO/EventLoop.swift:
```diff
@@ -222,6 +291,66 @@ extension EventLoop {
+    ///     - task: The closure that will be executed.
+    /// - return: `RepeatedTask`
+    @discardableResult
+    public func scheduleTaskWithFixedDelay(initialDelay: TimeAmount, delay: TimeAmount, _ task: @escaping () throws -> EventLoopFuture<Void>) -> RepeatedTask {
```
we need a test for this

weissi commented on Sources/NIO/EventLoop.swift:
```diff
@@ -222,6 +291,66 @@ extension EventLoop {
+    ///     - task: The closure that will be executed.
+    /// - return: `RepeatedTask`
+    @discardableResult
+    public func scheduleTaskWithFixedDelay(initialDelay: TimeAmount, delay: TimeAmount, _ task: @escaping () throws -> Void) -> RepeatedTask {
```
we need a test for this

weissi commented on Sources/NIO/EventLoop.swift:
```diff
@@ -222,6 +291,66 @@ extension EventLoop {
+    ///     - task: The closure that will be executed.
+    /// - return: `RepeatedTask`
+    @discardableResult
+    public func scheduleTaskAtFixedRate(initialDelay: TimeAmount, period: TimeAmount, _ task: @escaping () throws -> EventLoopFuture<Void>) -> RepeatedTask {
```
- we need a test for this
- this shouldn't be `throws` (as it's returning an `EventLoopFuture`)

weissi commented on Sources/NIO/EventLoop.swift:
```diff
@@ -222,6 +291,66 @@ extension EventLoop {
+    @discardableResult
+    public func scheduleTaskAtFixedRate(initialDelay: TimeAmount, period: TimeAmount, _ task: @escaping () throws -> Void) -> RepeatedTask {
+        let futureTask: () throws -> EventLoopFuture<Void> = {
+            try task()
```
- this should catch the error and make it a failed future

FranzBusch commented:
@weissi Thanks, for clearing that up! This makes sense for work which succeeds the promise immediately but in both of the implemented methods that cannot be assumed. Furthermore, `execute` is just an immediate scheduling mechanism and my intention with the implementation was to provide a convenient way to schedule any work repeating work with a delay and a rate/delay mode so that users don't have to implement that behaviour all on their own. Of course, you guys still decide if one of these methods actually endangers overloading the EL and thus should not be provided.
Or was your intention to actually change the implementation side of the repeated scheduling?

FranzBusch commented:
Just saw your Code example and I get what you mean. Yes your solution should be equal to mine :)

FranzBusch commented:
@weissi Just updated the PR and removed the fixed rate scheduling and added some tests. Is there something else I should test for?

About the fixed rate scheduling why does netty have the possibility and swift-nio shouldn't? Can't the ELs in netty get overloaded as well?

weissi commented on Sources/NIO/EventLoop.swift:
```diff
@@ -50,6 +50,61 @@ public struct Scheduled<T> {
+    /// Whether the execution of the task is immediately canceled depends on whether the execution of a task has already begun.
+    ///  This means immediate cancellation is not guaranteed.
+    public func cancel() {
+        scheduled?.cancel()
```
please prefix those (and the following) by `self.` (won't repeat this comment for all the other methods :) )

weissi commented on Tests/NIOTests/EventLoopTest.swift:
```diff
@@ -91,6 +91,64 @@ public class EventLoopTest : XCTestCase {
+    public func testScheduleRepeatedTask() throws {
+        let nanos = DispatchTime.now().uptimeNanoseconds
+        let initialDelay: TimeAmount = .milliseconds(50)
+        let delay: TimeAmount = .milliseconds(100)
```
can we maybe make this `5` and `10` milliseconds? Otherwise this test alone will run almost 1s which is quite slow

FranzBusch commented:
@weissi Done :)

weissi commented on Sources/NIO/EventLoop.swift:
```diff
@@ -50,6 +50,61 @@ public struct Scheduled<T> {
     }
 }
 
+public class RepeatedTask {
```
this needs to be `final`

weissi replied:
and have doc comments

weissi commented on Sources/NIO/EventLoop.swift:
```diff
@@ -50,6 +50,61 @@ public struct Scheduled<T> {
 
+public class RepeatedTask {
+    private let delay: TimeAmount
+    private var scheduled: Scheduled<EventLoopFuture<Void>>? {
```
oh sorry, didn't notice that both of those are `var`. Did you run the tests with [TSan (thread sanitizer)](https://www.shinobicontrols.com/wp-content/uploads/2016/08/Enable_Sanitizer.png) enabled? I'm sure TSan will find the threading error here.

The assumption (and that should actually be documented) is that the user can `cancel` a `RepeatedTask` from any thread but this isn't the case here.

weissi commented on Sources/NIO/EventLoop.swift:
```diff
@@ -50,6 +50,61 @@ public struct Scheduled<T> {
+    }
+
+    private func reschedule0() {
+        guard let task = task else {
```
`self.` missing

weissi commented on Sources/NIO/EventLoop.swift:
```diff
@@ -50,6 +50,61 @@ public struct Scheduled<T> {
+        guard let task = task else {
+            return
+        }
+        self.scheduled = self.scheduled?.futureResult.eventLoop.scheduleTask(in: delay, task)
```
`self.` missing

weissi commented on Sources/NIO/EventLoop.swift:
```diff
@@ -50,6 +50,61 @@ public struct Scheduled<T> {
+            if case EventLoopError.shutdown = error {
+                return
+            }
+            self.reschedule0()
```
what other errors could we hit here?

FranzBusch replied:
There could be an EventLoopError.unsupportedOperation, EventLoopError.shutdownFailed or any of the users thrown errors right? Should I handle the two EventLoopError cases?

Lukasa replied:
I don't think it can. If you `schedule` a function that returns a `Future`, the user thrown errors from that function cannot appear in the outer future.

Lukasa replied:
I think it may be fair to say that if the scheduled `futureResult` fails, there is no reason to reschedule, right?

FranzBusch replied:
Yes with the current status the only way the `futureResult` fails is either during shutdown with `.shutdown` or if the task is cancelled with `cancelled`. No other error is possible at the moment. So I can remove this `whenFailure` case.

weissi replied:
if you want to be sure it's one of those errors you could just assert that it's one of them:

    assert(([EventLoopError.cancelled, EventLoopError.shutdown] as [Error]).contains(error), "unexpected error: \(error)")

FranzBusch commented:
@weissi I ran the tests with TSan and noticed multiple data races. One was just an error in the tests and I used an Atomic counter. The other was because of a cancel on a different thread as you described. To fix this I introduced a lock in the `RepeatedTask`. Could you please take a look if the synchronization looks good?

FranzBusch commented on Tests/NIOTests/EventLoopTest.swift:
```diff
@@ -91,6 +91,89 @@ public class EventLoopTest : XCTestCase {
+        XCTAssertNoThrow(try eventLoopGroup?.syncShutdownGracefully())
+        XCTAssertNotNil(weakEventLoop)
+        eventLoopGroup = nil
+        Thread.sleep(forTimeInterval: 0.01)
```
I had to introduce a sleep here otherwise this test was really flaky and sometimes the `weakEventLoop` wasn't nil. Do you have any idea why that could be the case?

Lukasa commented on Sources/NIO/EventLoop.swift:
```diff
@@ -50,6 +50,67 @@ public struct Scheduled<T> {
+    private let delay: TimeAmount
+    private var scheduled: Scheduled<EventLoopFuture<Void>>?
+    private var task: (() throws -> EventLoopFuture<Void>)?
+    private let lock = Lock()
```
I don't think we want to use a lock here: instead, we should use the event loop that this `RepeatedTask` belongs to to manage synchronization.

The TL;DR for this is that locking is hard, and right now this isn't handling it very well. In particular, `begin` does not handle the synchronisation correctly. The other advantage is that with using the parent event loop we can easily check whether we are already holding the "lock" and so provide a fast-path that does not involve cross-loop dispatch.

FranzBusch replied:
I was thinking about the `begin` method as well but since this is only used internally and the user can only call cancel after `reschedule` was triggered once it should be fine. 

Would synchronization over the EL work with `executing` begin, reschedule and cancel on the EL? @Lukasa

Lukasa replied:
Yup, should work fine. In fact, only `cancel` should require the explicit synchronisation: both `begin` and `cancel` should by definition always be called on the loop, and so can just degenerately `assert` that that's what's actually happening.

Lukasa commented on Sources/NIO/EventLoop.swift:
```diff
@@ -50,6 +50,87 @@ public struct Scheduled<T> {
+        }
+    }
+
+    private func begin0(in delay: TimeAmount) {
```
This function should assert it's in the event loop.

weissi replied:
I'm deleting this comment as it's done.

Lukasa commented on Sources/NIO/EventLoop.swift:
```diff
@@ -50,6 +50,87 @@ public struct Scheduled<T> {
+    }
+
+    private func cancel0() {
+        assert(eventLoop.inEventLoop)
```
Nit: `self`

Lukasa commented on Sources/NIO/EventLoop.swift:
```diff
@@ -50,6 +50,87 @@ public struct Scheduled<T> {
+    }
+
+    private func reschedule() {
+        assert(eventLoop.inEventLoop)
```
Nit: `self`

Lukasa replied:
Nit: We could `guard let` here to clean up the unwrapping code a bit.

Lukasa commented on Sources/NIO/EventLoop.swift:
```diff
@@ -50,6 +50,87 @@ public struct Scheduled<T> {
+    }
+
+    private func reschedule0() {
+        assert(eventLoop.inEventLoop)
```
Nit: `self`

Lukasa commented on Sources/NIO/EventLoop.swift:
```diff
@@ -50,6 +50,87 @@ public struct Scheduled<T> {
+
+    private func reschedule0() {
+        assert(eventLoop.inEventLoop)
+        guard let task = self.task else {
```
May as well also unwrap `self.scheduled` here too.

weissi commented on Sources/NIO/EventLoop.swift:
```diff
@@ -396,7 +515,10 @@ internal final class SelectableEventLoop: EventLoop {
 
-        let scheduled = Scheduled(promise: promise, cancellationTask: {
+        let scheduled = Scheduled(promise: promise, cancellationTask: { [weak self] in
+            guard let `self` = self else {
```
we don't normally use weak references because their impact on performance. Why is this one necessary?

Lukasa commented on Sources/NIO/EventLoop.swift:
```diff
@@ -50,6 +50,83 @@ public struct Scheduled<T> {
+        guard let task = self.task, let scheduled = self.scheduled else {
+            return
+        }
+        self.scheduled = scheduled.futureResult.eventLoop.scheduleTask(in: self.delay, task)
```
Ah, we actually don't need the conditional binding of `scheduled` anymore, because we have an event loop on this object we can use.

Lukasa commented on Sources/NIO/EventLoop.swift:
```diff
@@ -93,7 +170,6 @@ public protocol EventLoop: EventLoopGroup {
     /// Schedule a `task` that is executed by this `SelectableEventLoop` after the given amount of time.
     func scheduleTask<T>(in: TimeAmount, _ task: @escaping () throws -> T) -> Scheduled<T>
 }
-
```
Nit: replace this whitespace line.

FranzBusch replied:
Can I leave this since it follows the overall formatting where no empty line is before the closing brackets?

Lukasa replied:
This line is after the closing brace.

Lukasa commented on Sources/NIO/EventLoop.swift:
```diff
@@ -50,6 +50,83 @@ public struct Scheduled<T> {
+    private let delay: TimeAmount
+    private let eventLoop: EventLoop
+    private var scheduled: Scheduled<EventLoopFuture<Void>>?
+    private var task: (() throws -> EventLoopFuture<Void>)?
```
I think we removed the need for this to be `throws`, didn't we? It doesn't actually manifest in the word `try` anywhere in the code for this object, but let's constrain the type further and not allow it to throw.

weissi commented:
@swift-nio-bot test this please

Lukasa commented:
Looks like `testScheduleRepeatedTask` is a bit racy and sometimes fails.

weissi commented:
@FranzBusch ok, we need to address the test flakiness first. I'd recommend to test it with a `for _ in 0..<10000 { /* real test code here */ }` and see if you can repro what our CI hit.

FranzBusch commented:
I tried running the tests in a loop but couldn't reproduce the flakiness. I just increased the timeouts a bit maybe they were a bit too optimistic.

Lukasa commented:
You may want to go to more than a second. The timeout should only be hit in failing cases anyway.

Lukasa commented on Tests/NIOTests/EventLoopTest.swift:
```diff
@@ -91,6 +91,88 @@ public class EventLoopTest : XCTestCase {
+        let expect = expectation(description: "Is cancelling RepatedTask")
+        var repeatedTask: RepeatedTask?
+        repeatedTask = eventLoopGroup.next().scheduleRepeatedTask(initialDelay: initialDelay, delay: delay) { () -> Void in
+            DispatchQueue.main.async {
```
Wait, what is going on here? You can't dispatch to the main dispatch queue.

Lukasa replied:
I'm also not really sure what the utility of doing that is. Just create a separate serial dispatch queue to use. :smile:

Lukasa commented:
@swift-nio-bot test this please

weissi commented:
@Lukasa given that @FranzBusch has filled the PR description out with the commit template, we could just use the 'squash & merge' button and it'll have the same effect, right?

Lukasa commented:
@weissi That was my original plan, but actually the PR description contains only 1/3 of the commit template, which is why I didn't land on that plan.

Lukasa commented:
@swift-nio-bot test this please

weissi commented:
@Lukasa oh sorry, didn't notice that.
